### ISSUE TITLE
Release V1

### ISSUE THREAD

**Issue by maennchen** [maennchen/ZipStream-PHP#71]
I've just released the version `v1.0.0-alpha.1`.

Please test the library and report any bugs you find.

**Comment by paciliojoe** [comment]
I am just getting started with ZipStream-PHP, so the issue may be on my end, but I am unable to zip anything other than a txt file.  I am able to zip one or multiple txt files, but once I try to add a jpg, png, or pdf, I am unable to open the zip archive that gets streamed to my browser.

(I also had to include a HttpHeaderCallback in order for the zip to stream to my browser in the first place - the default settings did not work)

 $zipOptions = new Archive();
        $zipOptions->setSendHttpHeaders(true);
        $zipOptions->setHttpHeaderCallback('header');

        $zip = new ZipStream('test.zip', $zipOptions);

        $root = \Yii::getAlias('@app').'/uploads/';

        $zip->addFile('test-file.txt', file_get_contents($root.'Test1.txt'));
        $zip->addFileFromPath('some_image.txt', $root.'Test1.txt');
        $zip->addFileFromPath('some_image.txt', $root.'Test2.txt');

        $zip->addFileFromPath('test-file.png', $root.'Image 010.png');  // Everything works without this line.  With this line, however the zip won't open.

        $zip->finish();

**Comment by paciliojoe** [comment]
I did some additional troubleshooting and it looks like it may be a contents length issue.  I tried to just output the jpg contents into a text file and it doesn't work unless I cut down the contents (see below).  This may be a red herring, but it seems consistent, as I get the same error if I create a LARGE text file and attempt to zip that.  

This works:
```
        $jpg = file_get_contents($root.'iStock-523790729.jpg');
        $txt = file_get_contents($root.'Test1.txt');

        $zip->addFile('test-file.txt', file_get_contents($root.'Test1.txt'));
        $zip->addFile('test-file2.txt', substr($jpg, 0, 10));
```
This does not:
```
        $jpg = file_get_contents($root.'iStock-523790729.jpg');
        $txt = file_get_contents($root.'Test1.txt');

        $zip->addFile('test-file.txt', file_get_contents($root.'Test1.txt'));
        $zip->addFile('test-file2.txt', $jpg);
```
Would this have something to do with a local PHP setting or something?

**Comment by maennchen** [comment]
@paciliojoe Thanks for testing the beta. I’ll try to reproduce the issue and then release beta 2.

**Comment by paciliojoe** [comment]
It looks like the header defaults are the only issue that I found with the beta.  I'm having the same jpg issues with previous versions.  I'll open a separate thread to track that (don't want to sidetrack the beta work)  

Thanks!

**Comment by volkeransmann** [comment]
I did some testing with the new version in regards of large file support in conjunction with S3 Streams.
I can confirm what @paciliojoe found out: without the setSendHttpHeaders and setHttpHeaderCallback Options set - the output is simply displayed in the browser. 
I also had to set the setZeroHeader to true. Otherwise the Browser would start downloading the zip and only after that prompt the download-dialog. (and also: the resulting zip would only be a few bytes and corrupt.)

That said: Zips are successfully created with S3 Sources!
But i discovered two Problems:
1. Archives above 2GB (could not pinpoint the exact size) are broken. (Tried adding a few 500MB Files, 1500MB was fine, 2500MB was corrupt. Max Archive-Size was always around 2,28GB)
2. The Native OSX-Unzipper can not handle the Archives. ("The Unarchiver" works fine, v0.5.2 worked also with the Native OSX-Unzipper)

Maybe my implementation is wrong? 

```
$keys = array(
	'name.mp4' => 'key-in-s3',
	'name2.mp4' => 'key-in-s3',
	'name3.mp4' => 'key-in-s3',
);

// Get Laravel Disk Instance, Adapter, Client etc...
$bucket-name = 'bucket-name';
$disk = Storage::disk($region);
$adapter = $disk->getAdapter();
$client = $adapter->getClient();
$client->registerStreamWrapper();

return new StreamedResponse(function() use($client, $keys, $bucket-name) {

	$options = new ArchiveOptions();
	$options->setZeroHeader(true); 
	$options->setEnableZip64(true);
	$options->setSendHttpHeaders(true);
	$options->setHttpHeaderCallback('header');

	# create a new zipstream object
	$zip = new ZipStream('example.zip', $options);

	foreach ($keys as $name => $key) {

		// Open a stream in read-only mode
		if (!($stream = fopen("s3://{$bucket-name}/{$key}", 'r'))) {
			throw new \Exception('Could not open stream for reading file: ['.$key.']');
		}

		$fileOptions = new FileOptions();
		$fileOptions->setMethod(Method::STORE());
		$zip->addFileFromStream($name, $stream, $fileOptions);

	}

	# finish the zip stream
	$zip->finish();

});
```

**Comment by yanchoo** [comment]
> Archives above 2GB (could not pinpoint the exact size) are broken. (Tried adding a few 500MB Files, 1500MB was fine, 2500MB was corrupt. Max Archive-Size was always around 2,28GB)

I had a similar issue then I opened error.log and s\

`PHP Fatal error:  Maximum execution time of 30 seconds exceeded in  ...maennchen/zipstream-php/src/ZipStream.php on line 649`

**Comment by volkeransmann** [comment]
Thanks! It's working now as expected. (but still not with the native osx extractor)
Although i have to deal with possible rogue scripts now. But that can be dealt with.

**Comment by ctf0** [comment]
@volkeransmann disable `$options->setZeroHeader(true); ` and it will work with osx archiver `10.11.6`

`$options->setEnableZip64(true);` not needed as its enabled by default.

**Comment by volkeransmann** [comment]
@ctf0 as i already wrote: disabling setZeroHeader results in currupt files or "504 Gateway Time-out
"-Errors (max_execution_time set to zero). And the browser starts downloading the zip without prompting the "save-dialog". So if the file is small enough it will download right away and only after finishing - it will prompt the user to select a destination. That is not a desired outcome because it leaves the user clueless. Even if it would work for large files. Which it does not.

Maybe your solution works for normal files and just not in conjunction with "addFileFromStream"

**Comment by volkeransmann** [comment]
also using laravel. I am talking about archive sizes > 2GB

Do you also add files from S3 - or from local storage?

**Comment by ctf0** [comment]
i will do some tests with the archive and report back.

u mean the manager ? 
- if yes then it works with whatever disk u have in ur laravel, but i havent done much testing on cloud disks because of my crappy connection, any help is appreciated though :nerd_face: 

- if no then i will be using local storage

**Comment by volkeransmann** [comment]
don't know about a manager.
But there seems to be a difference in using $zip->addFile and $zip->addFileFromStream.
I think the later requires the ZeroHeader because ZIpStream has no knowledge of the final size because of the nature of streamed responses.

Just try to add a large file from S3 with my code from above and you can witness in your network monitor that data is beeing transfered and only after that is the download-dialog presented. (Or an Error thrown for larger files...)

either way. Not that much of a problem. Relatively easy to explain to osx-users not to use the native extractor.

**Comment by ctf0** [comment]
in the meanwhile can u try changing ur code to this and try again "edit what u need"

```php
return response()->stream(function () use ($name, $list) {
    
    $options = new Archive();
    $options->setContentType('application/octet-stream');
    $options->setSendHttpHeaders(true);
    $options->setHttpHeaderCallback('header');
    $options->setDeflateLevel(9);

    $zip = new ZipStream("$name.zip", $options);

    foreach ($list as $file) {
        $file_name = $file['name'];
        $streamRead = Storage::readStream($file));

        if ($streamRead) {
            $zip->addFileFromStream($file_name, $streamRead);
        } else {
            logger('error');
        }
    }
    
    $zip->finish();
});
```

this is better because
- `fopen` doesnt work with https
- u stream the zip as its being built instead of holding it

**Comment by ctf0** [comment]
@volkeransmann tested on local zipping `2.11 gb` 
- `$options->setZeroHeader(false);` no probs
- `$options->setZeroHeader(true);` produces `name.zip.cpgz` which unzipping it will simply keep producing other zipped files

**Comment by volkeransmann** [comment]
@ctf0 Again - i assume you did not test with files hosted on S3. The behaviour is different there.
I just used your code - and while the resulting zip is indeed extracteable with OSX Native Extractor - the "Save-Under"-Dialog is still only presented to the user AFTER the file is completely downloaded.
So no user-feedback meanwhile. This is not usable in Production i am afraid...

While my code from the initial comment here is not working with the native extractor - it fullfills every other need. It also starts streaming/downloading right away without storing on webserver and it presents the save-under-dialog right away.

Just try yourself with any S3 Disk on your code:

`$disk` = Storage::disk('your-s3-disk-here');

		return response()->stream(function () use ($keys, $disk) {

			$options = new ArchiveOptions();
			$options->setContentType('application/octet-stream');
			$options->setSendHttpHeaders(true);
			$options->setHttpHeaderCallback('header');
			$options->setDeflateLevel(9);

			$zip = new ZipStream('example.zip', $options);

			foreach ($keys as $name => $s3key) {

				$streamRead = $disk->readStream($s3key);

				if ($streamRead) {
					$zip->addFileFromStream($name, $streamRead);
				} else {
					logger('error');
				}
			}

			$zip->finish();
		});
`

**Comment by ctf0** [comment]
okay good, so half of the problem is solved.

for the notification of the zipping, i would recommend using websockets to stream the progress back to the user [example](https://github.com/ctf0/Laravel-Media-Manager/blob/master/src/Controllers/Modules/Download.php#L55)

as for trying with s3, my connection isnt strong enough to download 2 gb file to test with, still hopefully with the ws notification trick ur issue would be solved for good 👍

**Comment by volkeransmann** [comment]
You don't have to try with large files. The issue with S3 appears also with small files. They should just have to be big enough to take some time to download on your connection. Otherwise you will not notice the difference between:

A) Finish Download -> Present Dialog (User is left clueless during download)
B) Present Dialog -> Start Download (normal Browser behaviour)

Although i like the idea of having the possibility to present a progress (i would extend that to display actual progress based on size not on file-count) i am not sure that i am willing to introduce the complexity. A Download should behave consistent and make use of the built-in browser mechanism.
Also i am not sure how your workaround would work in reality. Most of the time a download is triggered to open in a new window (most modern browsers detect the download there an close this window immediately after the download-dialog is closed). But because this dialog is not presented before but after the download - i suspect the new tab would stay open. So Users have to manually go back to the initial tab to see the websocket progress.

Your solution feels like a workaround to me. Also not practical from a user standpoint i assume. I'll just stick to my solution and tell users beforehand or via FAQ not to use the native OSX Extractor.

But thanks for getting involved! Maybe you'll find a better solution than me once you test with an actual S3 Bucket. (i think key is to use the StreamWrapper from S3: https://aws.amazon.com/de/blogs/developer/amazon-s3-php-stream-wrapper/)

**Comment by aprat84** [comment]
I'm testing it cause the 4GB file limit for 32 bit zip, and for now everything is ok. Downloading 5GB file with no issues.

I've noticed there's a new option to set deflate level, but it is not used when adding small files, if I'm not mistaken.
Shouldn't this code ([src/File.php:131](https://github.com/maennchen/ZipStream-PHP/blob/e03c18f819fc35c6f190cecef9f9ec80e5fd662c/src/File.php#L131))
```php
        // compress data if needed
        if ($this->method->equals(Method::DEFLATE())) {
            $data = gzdeflate($data);
        }
```
Be like this?
```php
        // compress data if needed
        if ($this->method->equals(Method::DEFLATE())) {
            $data = gzdeflate($data, $this->opt->getDeflateLevel());
        }
```

**Comment by NicolasCARPi** [comment]
@maennchen I believe now would be a good time to release v1.0.0-alpha.2, don't you think?

Can you do it please?

**Comment by volkeransmann** [comment]
Hi @maennchen ,

i tested the newest release and while it works perfectly in conjunction with S3-Files (even 4Gb+) all
zips created throw an error in the osx native "Archive Utility". However the terminal command "unzip" extracts the archive just fine.

Before i implement a user-warning to not use the native osx extractor: Is there any chance we can tacle this issue and come up with a workaround?

I noticed that a few libraries are using .tar for OSX-Clients. But i am afraid this will introduce even more confusion on the user side. Also Chrome seems to warn about .tar downloads nowadays...
What i do know is that this is fixable in general. Just tried downloading a 10Gb Zip from  One-Drive Online and the resulting zip opens just fine with the native osx extractor.

I read a lot about the osx extractor and know that the underlying issue is that this program just does not respect the zip standard. But a solution or workaround would be great. 

Let me know if you need examples or logs. 

thanks,
volker

**Comment by NicolasCARPi** [comment]
@volkeransmann see https://github.com/maennchen/ZipStream-PHP/issues/76

If problem is still there, please open a new issue with the code you are using.
